homeassistant/components/systemmonitor/sensor.py

```diff
@@ -487,6 +501,29 @@ def _update(  # noqa: C901
         state = round(_disk_usage(data.argument).used / 1024**3, 1)
     elif type_ == "disk_free":
         state = round(_disk_usage(data.argument).free / 1024**3, 1)
+    elif type_ == "disk_await":
+        state = None
+        counters = None
+        if data.argument:
+            all_counters = _disk_io_counters(perdisk=True)
+            if data.argument in all_counters:
+                counters = all_counters[data.argument]
+            else:
+                _LOGGER.warning(
+                    "Disk %s not found in: %s", data.argument, all_counters.keys()
+                )
+        else:
+            counters = _disk_io_counters(perdisk=False)
+        if counters and data.value:
+            iop_time_ms = (counters.read_time + counters.write_time) - (
+                data.value.read_time + data.value.write_time
+            )
+            num_iops = (counters.read_count + counters.write_count) - (
+                data.value.read_count + data.value.write_count
+            )
+            if num_iops > 0:
+                state = round(iop_time_ms / num_iops, 1)
+        value = counters
```

Comment:
This is not a value provided by the device or service but a calculated value. We want to avoid adding such calculations in Home Assistant.

../Frenck

Reply:
@frenck this is the same math that is already happening below to calculate `throughput_network_in` and `throughput_network_out`.

Do you suggest that the component instead log `read_time`, `write_time`, `read_count` and `write_count`, and then leave it to the user to create template sensors that does the math?

If we subprocess `iostat -x` to get the `r_await` and `w_await` statistics instead of calling `psutil.disk_io_counters()`, does that make it count as a value provided by the device?

Almost all periodic statistics are calculated values based on monotonic counters that the kernel provides. We return `processor_use` as `psutil.cpu_percent()`, but that's provided by the following calculations that psutil makes:
```
def cpu_times():
    # read system counters for:
    #  - user
    #  - system
    #  - idle
    #  - nice (UNIX)
    #  - iowait (Linux)
    #  - irq (Linux, FreeBSD)
    #  - softirq (Linux)
    #  - steal (Linux >= 2.6.11)
    #  - guest (Linux >= 2.6.24)
    #  - guest_nice (Linux >= 3.2.0)

def _cpu_tot_time(times):
    tot = sum(times)
    tot -= getattr(times, "guest", 0)
    tot -= getattr(times, "guest_nice", 0)
    return tot

def _cpu_busy_time(times):
    busy = _cpu_tot_time(times)
    busy -= times.idle
    busy -= getattr(times, "iowait", 0)
    return busy

def _cpu_times_deltas(t1, t2):
    field_deltas = []
    for field in _psplatform.scputimes._fields:
        field_delta = getattr(t2, field) - getattr(t1, field)
        field_deltas.append(field_delta)
    return _psplatform.scputimes(*field_deltas)

def calculate(t1, t2):
    times_delta = _cpu_times_deltas(t1, t2)
    all_delta = _cpu_tot_time(times_delta)
    busy_delta = _cpu_busy_time(times_delta)
    busy_perc = (busy_delta / all_delta) * 100

def cpu_percent():
    t1 = _last_cpu_times
    _last_cpu_times = cpu_times()
    return calculate(t1, _last_cpu_times)
```

Reply:
> this is the same math that is already happening below to calculate

Just because we did these things in the past, doesn't mean we still allow these things. Not all the things we did back then turned out to be good ideas.

../Frenck

Reply:
Do you propose that calculations be moved up into user templates, or down into psutil?